File: juno_rs/src/take_profit/trending.rs
use super::TakeProfit;
use crate::{
    easing::{tween, Easing, StdRngExt},
    genetics::Chromosome,
    indicators::Adx,
    math::lerp,
    Candle,
};
use juno_derive_rs::*;
use rand::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Chromosome, Clone, Debug, Deserialize, Serialize)]
pub struct TrendingParams {
    pub up_thresholds: (f64, f64),
    pub down_thresholds: (f64, f64),
    pub period: u32,
    pub lock_threshold: bool,
    pub easing: Easing,
}

fn up_thresholds(rng: &mut StdRng) -> (f64, f64) {
    loop {
        let (s, l) = (rng.gen_range(0.001..9.999), rng.gen_range(0.002..10.000));
        if s < l {
            return (s, l);
        }
    }
}
fn down_thresholds(rng: &mut StdRng) -> (f64, f64) {
    up_thresholds(rng)
}
fn period(rng: &mut StdRng) -> u32 {
    rng.gen_range(1..200)
}
fn lock_threshold(rng: &mut StdRng) -> bool {
    rng.gen_bool(0.5)
}
fn easing(rng: &mut StdRng) -> Easing {
    rng.gen_easing()
}

pub struct Trending {
    up_min_threshold: f64,
    up_max_threshold: f64,
    down_min_threshold: f64,
    down_max_threshold: f64,
    lock_threshold: bool,
    easing: Easing,
    up_threshold_factor: f64,
    down_threshold_factor: f64,
    adx: Adx,
    close_at_position: f64,
    close: f64,
}

impl Trending {
    fn set_threshold_factors(&mut self) {
        let adx_value = self.adx.value / 100.0;
        let progress = tween(adx_value, self.easing);
        let up_threshold = lerp(self.up_min_threshold, self.up_max_threshold, progress);
        let down_threshold = lerp(self.down_min_threshold, self.down_max_threshold, progress);
        self.up_threshold_factor = 1.0 + up_threshold;
        self.down_threshold_factor = 1.0 - down_threshold;
    }
}

impl TakeProfit for Trending {
    type Params = TrendingParams;

    fn new(params: &Self::Params) -> Self {
        Self {
            up_min_threshold: params.up_thresholds.0,
            up_max_threshold: params.up_thresholds.1,
            down_min_threshold: params.down_thresholds.0,
            down_max_threshold: params.down_thresholds.1,
            lock_threshold: params.lock_threshold,
            easing: params.easing,
            up_threshold_factor: 0.0,
            down_threshold_factor: 0.0,
            adx: Adx::new(params.period),
            close_at_position: 0.0,
            close: 0.0,
        }
    }

    fn upside_hit(&self) -> bool {
        self.close >= self.close_at_position * self.up_threshold_factor
    }

    fn downside_hit(&self) -> bool {
        self.close <= self.close_at_position * self.down_threshold_factor
    }

    fn clear(&mut self, candle: &Candle) {
        self.close_at_position = candle.close;
        if self.lock_threshold {
            self.set_threshold_factors();
        }
    }

    fn update(&mut self, candle: &Candle) {
        self.close = candle.close;
        self.adx.update(candle.high, candle.low);
        if !self.lock_threshold {
            self.set_threshold_factors();
        }
    }
}

File: juno/take_profit/trending.py
from decimal import Decimal
from typing import Callable, Optional

import pytweening

from juno import Candle
from juno.indicators import Adx
from juno.math import lerp

from .take_profit import TakeProfit

_EASINGS = {
    "linear": pytweening.linear,
    "quad_in": pytweening.easeInQuad,
    "quad_out": pytweening.easeOutQuad,
    "quad_inout": pytweening.easeInOutQuad,
    "cubic_in": pytweening.easeInCubic,
    "cubic_out": pytweening.easeOutCubic,
    "cubic_inout": pytweening.easeInOutCubic,
    "quart_in": pytweening.easeInQuart,
    "quart_out": pytweening.easeOutQuart,
    "quart_inout": pytweening.easeInOutQuart,
    "quint_in": pytweening.easeInQuint,
    "quint_out": pytweening.easeOutQuint,
    "quint_inout": pytweening.easeInOutQuint,
    "sine_in": pytweening.easeInSine,
    "sine_out": pytweening.easeOutSine,
    "sine_inout": pytweening.easeInOutSine,
    "circ_in": pytweening.easeInCirc,
    "circ_out": pytweening.easeOutCirc,
    "circ_inout": pytweening.easeInOutCirc,
    "expo_in": pytweening.easeInExpo,
    "expo_out": pytweening.easeOutExpo,
    "expo_inout": pytweening.easeInOutExpo,
    "elastic_in": pytweening.easeInElastic,
    "elastic_out": pytweening.easeOutElastic,
    "elastic_inout": pytweening.easeInOutElastic,
    "back_in": pytweening.easeInBack,
    "back_out": pytweening.easeOutBack,
    "back_inout": pytweening.easeInOutBack,
    "bounce_in": pytweening.easeInBounce,
    "bounce_out": pytweening.easeOutBounce,
    "bounce_inout": pytweening.easeInOutBounce,
}


class Trending(TakeProfit):
    _up_min_threshold: Decimal
    _up_max_threshold: Decimal
    _down_min_threshold: Decimal
    _down_max_threshold: Decimal
    _lock_threshold: bool
    _easing: str
    _up_threshold_factor: Decimal = Decimal("0.0")
    _down_threshold_factor: Decimal = Decimal("0.0")
    _adx: Adx
    _close_at_position: Decimal = Decimal("0.0")
    _close: Decimal = Decimal("0.0")

    def __init__(
        self,
        up_thresholds: tuple[Decimal, Decimal],
        down_thresholds: Optional[tuple[Decimal, Decimal]] = None,
        period: int = 14,
        lock_threshold: bool = False,
        easing: str = "linear",
    ) -> None:
        if down_thresholds is None:
            down_thresholds = up_thresholds
        assert 0 <= up_thresholds[0] and 0 <= up_thresholds[1]
        assert 0 <= down_thresholds[0] and 0 <= down_thresholds[1]
        self._up_min_threshold = up_thresholds[0]
        self._up_max_threshold = up_thresholds[1]
        self._down_min_threshold = down_thresholds[0]
        self._down_max_threshold = down_thresholds[1]
        self._lock_threshold = lock_threshold
        self._adx = Adx(period)
        self._easing = easing

    @property
    def upside_hit(self) -> bool:
        return self._close >= self._close_at_position * self._up_threshold_factor

    @property
    def downside_hit(self) -> bool:
        return self._close <= self._close_at_position * self._down_threshold_factor

    def clear(self, candle: Candle) -> None:
        self._close_at_position = candle.close
        if self._lock_threshold:
            self._set_thresholds()

    def update(self, candle: Candle) -> None:
        self._close = candle.close
        self._adx.update(candle.high, candle.low)
        if not self._lock_threshold:
            self._set_thresholds()

    def _set_thresholds(self) -> None:
        # The ADX value is essentially a progress from 0..1. Hence we can apply easing function
        # directly on it.
        adx_value = self._adx.value / 100
        progress = self._ease(adx_value)
        up_threshold = lerp(self._up_min_threshold, self._up_max_threshold, progress)
        down_threshold = lerp(self._down_min_threshold, self._down_max_threshold, progress)
        self._up_threshold_factor = 1 + up_threshold
        self._down_threshold_factor = 1 - down_threshold

    @property
    def _ease(self) -> Callable[[Decimal], Decimal]:
        easing_fn = _EASINGS.get(self._easing)
        if easing_fn is None:
            raise ValueError(f"Unknown easing function: {self._easing}")
        return easing_fn
